api/v1alpha1/etcdcluster_types.go

```diff
@@ -56,6 +56,7 @@ type EtcdClusterStatus struct {
 	// (as we may have just created a peer resource that doesn't have a pod yet, or the pod could be restarting), and it
 	// doesn't mean the number of members the etcd cluster has live, as pods may not be ready yet or network problems
 	// may mean the cluster has lost a member.
+	// +optional
 	Replicas int32 `json:"replicas"`
```

Comment:
If you're making this optional, then it should be a pointer?

Reply:
Yeah, I did consider it but I saw that DeploymentStatus.Replicas is an optional `int32`and copied that.
 * https://github.com/kubernetes/kubernetes/blob/v1.17.0/staging/src/k8s.io/api/apps/v1/types.go#L381

A counter example is CollisionCount, which is an `*int32`
 * https://github.com/kubernetes/kubernetes/blob/v1.17.0/staging/src/k8s.io/api/apps/v1/types.go#L408

This change shouldn't be necessary, but without it the Patch operation seemed to fail with an openapi validation error about "replicas being a required field" or some such.

I'll see if I can find out why that validation is being triggered and otherwise make this a pointer as you suggest.

Reply:
FYI This is the error, and note that this is not coming from the webhook validation, but from the K8S API server validation:

```
2020-01-07T12:04:53.698Z	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/mutate-etcd-improbable-io-v1alpha1-etcdcluster", "UID": "a5ef8c81-2d95-42b6-8171-e8fe250a9a5e", "kind": "etcd.improbable.io/v1alpha1, Kind=EtcdCluster", "resource": {"group":"etcd.improbable.io","version":"v1alpha1","resource":"etcdclusters"}}
2020-01-07T12:04:53.699Z	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/mutate-etcd-improbable-io-v1alpha1-etcdcluster", "UID": "a5ef8c81-2d95-42b6-8171-e8fe250a9a5e", "allowed": true, "result": {}, "resultError": "got runtime.Object without object metadata: &Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,RemainingItemCount:nil,},Status:,Message:,Reason:,Details:nil,Code:200,}"}
2020-01-07T12:04:53.702Z	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/validate-etcd-improbable-io-v1alpha1-etcdcluster", "UID": "2591c6e9-f98f-4d4e-b335-e9fb9a87b115", "kind": "etcd.improbable.io/v1alpha1, Kind=EtcdCluster", "resource": {"group":"etcd.improbable.io","version":"v1alpha1","resource":"etcdclusters"}}
2020-01-07T12:04:53.702Z	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/validate-etcd-improbable-io-v1alpha1-etcdcluster", "UID": "2591c6e9-f98f-4d4e-b335-e9fb9a87b115", "allowed": true, "result": {}, "resultError": "got runtime.Object without object metadata: &Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,RemainingItemCount:nil,},Status:,Message:,Reason:,Details:nil,Code:200,}"}
2020-01-07T12:04:53.734Z	DEBUG	controllers.EtcdCluster	Created Service	{"cluster": "teste2e-parallel-backup/e2e-backup-cluster", "service": "e2e-backup-cluster"}
2020-01-07T12:04:53.740Z	ERROR	controllers.EtcdCluster	Failed to update status	{"cluster": "teste2e-parallel-backup/e2e-backup-cluster", "error": "EtcdCluster.etcd.improbable.io \"e2e-backup-cluster\" is invalid: []: Invalid value: map[string]interface {}{\"apiVersion\":\"etcd.improbable.io/v1alpha1\", \"kind\":\"EtcdCluster\", \"metadata\":map[string]interface {}{\"annotations\":map[string]interface {}{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"etcd.improbable.io/v1alpha1\\\",\\\"kind\\\":\\\"EtcdCluster\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"name\\\":\\\"e2e-backup-cluster\\\",\\\"namespace\\\":\\\"teste2e-parallel-backup\\\"},\\\"spec\\\":{\\\"replicas\\\":1,\\\"storage\\\":{\\\"volumeClaimTemplate\\\":{\\\"resources\\\":{\\\"requests\\\":{\\\"storage\\\":\\\"100Mi\\\"}},\\\"storageClassName\\\":\\\"standard\\\"}}}}\\n\"}, \"creationTimestamp\":\"2020-01-07T12:04:53Z\", \"generation\":1, \"name\":\"e2e-backup-cluster\", \"namespace\":\"teste2e-parallel-backup\", \"resourceVersion\":\"1216\", \"uid\":\"d4c6a2fa-9c39-4b5b-ae25-f98a34f20ab4\"}, \"spec\":map[string]interface {}{\"replicas\":1, \"storage\":map[string]interface {}{\"volumeClaimTemplate\":map[string]interface {}{\"accessModes\":[]interface {}{\"ReadWriteOnce\"}, \"resources\":map[string]interface {}{\"requests\":map[string]interface {}{\"storage\":\"100Mi\"}}, \"storageClassName\":\"standard\", \"volumeMode\":\"Filesystem\"}}}, \"status\":map[string]interface {}{\"members\":[]interface {}{}}}: validation failure list:\nstatus.replicas in body is required"}

```

Reply:
`validation failure list:\nstatus.replicas in body is required`

Reply:
I've changed my mind. There's no point making this a pointer.
Adding the `// +optional` annotation allows it to be omitted in CREATE, UPDATE, and PATCH operations,
and in which case its zero-value `0` will be used....which makes sense...as that's what the the controller currently does anyway....sets it to the length of the `EtcdPeerList`.

You could argue that there should be a way to record in the status that the EtcdPeerList length is unknown, in which case it might make sense to turn this into a pointer, but that seems like a separate issue.

Other examples of this pattern are found in Cluster API: https://github.com/kubernetes-sigs/cluster-api/issues/1188

Reply:
This is correct IMO - the `// +optional` influences the OpenAPI spec, which is what is being used during the validation above. It hasn't come up previously because the lack of `omitempty` meant that when an Update was performed, the serialized JSON would contain `replicas: 0` - when patching, presumably the merge patch algorithm has pruned this field as it has not changed, hence you now get an error 😬